docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,387 @@
+|                     | `INT2`                  | Trivial / validated |
+|                     | `INT4`                  | Trivial / validated |
+|                     | `INTERVAL`              |                     |
+|                     | `STRING`                |                     |
+|                     | `TIME`                  |                     |
+|                     | `TIMESTAMP`             |                     |
+| `STRING(n)`         |                         |                     |
+|                     | `BOOL`                  |                     |
+|                     | `BYTES`                 | Trivial             |
+|                     | `DATE`                  |                     |
+|                     | `DECIMAL`               |                     |
+|                     | `FLOAT`                 |                     |
+|                     | `INET`                  |                     |
+|                     | `INT`                   |                     |
+|                     | `INTERVAL`              |                     |
+|                     | `JSONB`                 | Cast (extension)    |
+|                     | `STRING(n-1)`           | Validated           |
+|                     | `STRING(n+1)`           | Trivial             |
+|                     | `TIME`                  |                     |
+|                     | `TIMESTAMP`             |                     |
+| `STRING COLLATE ab` |                         |                     |
+|                     | `STRING COLLATE cd`     |                     |
+| `TIME(TZ)`          |                         |                     |
+|                     | `INTERVAL`              |                     |
+|                     | `STRING`                |                     |
+|                     | `TIME(TZ)`              | Trivial             |
+| `TIMESTAMP(TZ)`     |                         |                     |
+|                     | `INT`                   |                     |
+|                     | `DECIMAL`               |                     |
+|                     | `FLOAT`                 |                     |
+|                     | `TIME`                  |                     |
+|                     | `DATE`                  |                     |
+|                     | `STRING`                |                     |
+|                     | `TIMESTAMP(TZ)`         | Trivial             |
+| `UUID`              |                         |                     |
+|                     | `BYTES`                 | Trivial             |
+|                     | `STRING`                | Cast (extension)    |
+
+### Array conversions
+
+The element type of an array can be converted, based on the conversion
+rules above. That is, `STRING[]` could be converted to `DATE[]`.
+Additionally, it is possible to convert any type to an array of a
+convertable type, e.g. `STRING -> DATE[]`.
```

Comment:
can you expand on this. What is the expected semantics of this last conversion? What does pg does in that case? What is the use case?

In general I would be rather strongly against the idea of enabling an automatic (i.e. non-assisted) column type conversion that is not otherwise possible via a cast (using `::`). This last example would be an example of that. I'd need a use case to be convinced otherwise.